File: TwitchDVR/TwitchDVR/UpdateChecker.swift
import Foundation

struct UpdateInfo: Equatable {
    let version: String
    let tag: String
    let assetURL: URL
    let assetName: String
}

enum UpdateCheckState: Equatable {
    case idle
    case checking
    case upToDate(current: String, latest: String)
    case updateAvailable(UpdateInfo)
    case downloading(UpdateInfo)
    case error(String)
}

/// Fetches the latest GitHub release and downloads its packaged app.
struct UpdateChecker {
    static let repoOwner = "kamil004"
    static let repoName = "StreamDVR"
    static let assetPrefix = "StreamDVR-macOS"

    private static var latestReleaseURL: URL {
        URL(string: "https://api.github.com/repos/\(repoOwner)/\(repoName)/releases/latest")!
    }

    static func stripV(_ s: String) -> String {
        s.hasPrefix("v") ? String(s.dropFirst()) : s
    }

    /// 1.0.9 < 1.0.15 < 1.1.0
    static func compare(_ a: String, _ b: String) -> ComparisonResult {
        let av = a.components(separatedBy: ".").compactMap { Int($0) }
        let bv = b.components(separatedBy: ".").compactMap { Int($0) }
        for i in 0..<max(av.count, bv.count) {
            let x = i < av.count ? av[i] : 0
            let y = i < bv.count ? bv[i] : 0
            if x != y { return x < y ? .orderedAscending : .orderedDescending }
        }
        return .orderedSame
    }

    /// Latest release information for the packaged macOS asset, or nil when
    /// none exists yet / the request fails.
    static func fetchLatest() async -> UpdateInfo? {
        var request = URLRequest(url: latestReleaseURL)
        request.timeoutInterval = 20
        request.setValue("StreamDVR/1.0 (macOS)", forHTTPHeaderField: "User-Agent")
        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let http = response as? HTTPURLResponse, http.statusCode == 200,
              let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let tag = json["tag_name"] as? String,
              let assets = json["assets"] as? [[String: Any]] else {
            return nil
        }
        guard let asset = assets.first(where: {
            let name = $0["name"] as? String ?? ""
            return name.hasPrefix(assetPrefix) && name.hasSuffix(".zip")
        }),
        let assetName = asset["name"] as? String,
        let urlStr = asset["browser_download_url"] as? String,
        let url = URL(string: urlStr) else {
            return nil
        }
        return UpdateInfo(version: stripV(tag), tag: tag, assetURL: url, assetName: assetName)
    }

    /// Downloads the release zip to a temporary location.
    static func download(_ info: UpdateInfo) async throws -> URL {
        var request = URLRequest(url: info.assetURL)
        request.setValue("StreamDVR/1.0 (macOS)", forHTTPHeaderField: "User-Agent")
        let (zipURL, _) = try await URLSession.shared.download(for: request)
        return zipURL
    }
}

File: Windows/TwitchDVR.Core/UpdateChecker.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace TwitchDVR.Core;

public record UpdateInfo(string Version, string Tag, Uri AssetUrl, string AssetName);

public enum UpdateCheckState { Idle, Checking, UpToDate, UpdateAvailable, Downloading, Error }

public static class UpdateChecker
{
    public const string RepoOwner = "kamil004";
    public const string RepoName = "StreamDVR";
    public const string AssetPrefix = "TwitchDVR-Windows";

    static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(20) };

    public static string StripV(string s) => s.StartsWith('v') ? s[1..] : s;

    public static int CompareVersions(string a, string b)
    {
        var av = a.Split('.', StringSplitOptions.RemoveEmptyEntries).Select(s => int.TryParse(s, out var v) ? v : 0).ToArray();
        var bv = b.Split('.', StringSplitOptions.RemoveEmptyEntries).Select(s => int.TryParse(s, out var v) ? v : 0).ToArray();
        for (var i = 0; i < Math.Max(av.Length, bv.Length); i++)
        {
            var x = i < av.Length ? av[i] : 0;
            var y = i < bv.Length ? bv[i] : 0;
            if (x != y) return x.CompareTo(y);
        }
        return 0;
    }

    public static async Task<UpdateInfo?> FetchLatestAsync()
    {
        using var request = new HttpRequestMessage(HttpMethod.Get,
            $"https://api.github.com/repos/{RepoOwner}/{RepoName}/releases");
        request.Headers.TryAddWithoutValidation("User-Agent", "TwitchDVR/1.0 (Windows)");
        try
        {
            var response = await Http.SendAsync(request);
            if (!response.IsSuccessStatusCode) return null;
            var json = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return null;

            UpdateInfo? best = null;
            foreach (var release in doc.RootElement.EnumerateArray())
            {
                if (!release.TryGetProperty("assets", out var assets)) continue;
                var tag = release.TryGetProperty("tag_name", out var t) ? t.GetString() ?? "" : "";
                var published = release.TryGetProperty("published_at", out var p) && DateTime.TryParse(p.GetString(), out var dt) ? dt : DateTime.MinValue;

                foreach (var asset in assets.EnumerateArray())
                {
                    var name = asset.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                    if (!name.StartsWith(AssetPrefix, StringComparison.OrdinalIgnoreCase) || !name.EndsWith(".zip", StringComparison.OrdinalIgnoreCase)) continue;
                    var urlStr = asset.TryGetProperty("browser_download_url", out var u) ? u.GetString() : null;
                    if (string.IsNullOrEmpty(urlStr) || !Uri.TryCreate(urlStr, UriKind.Absolute, out var url)) continue;

                    var version = StripV(name.Replace(AssetPrefix, "").Replace("-", "").Replace(".zip", ""));
                    var info = new UpdateInfo(version, tag, url, name);
                    if (best == null || published > DateTime.MinValue)
                    {
                        best = info;
                        break;
                    }
                }
                if (best != null) break;
            }
            return best;
        }
        catch { return null; }
    }
}
